Confluent avro using https + user / password not working

**Logstash information**:

Please include the following information:

1. Logstash version: 7.13.3
2. Logstash installation source: Docker
3. How is Logstash being run: Docker
4. How was the Logstash Plugin installed:
Upgraded to latest stable using **logstash-plugin update logstash-integration-kafka** (v10.8.1)

**JVM** (e.g. `java -version`): Bundled JDK

**OS version** (`uname -a` if on a Unix-like system): 3.10.0-1160.24.1.el7.x86_64 #1 SMP Thu Mar 25 21:21:56 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**: 
Using HTTPS + auth does not work. Switching to HTTP with auth works however. 

**Steps to reproduce**:

```
input {
  kafka {
        bootstrap_servers => "mybroker.kafka:9094"
        group_id => "my_group"
        client_id => "my_client"
        topics => "mytopic"
        schema_registry_url => "https://my.registry:8081/"
        schema_registry_key => "myuser"
        schema_registry_secret => "mypassword"
        codec => "json"
        decorate_events => true
        auto_offset_reset => "earliest"
        jaas_path => "/path/to/kafka-jaas-input.conf"
        sasl_mechanism => "SCRAM-SHA-512"
        security_protocol => "SASL_SSL"
        ssl_truststore_location => "/path/to/certs/my.truststore.jks"
        ssl_truststore_password => "REDACTED"
        key_deserializer_class => "org.apache.kafka.common.serialization.ByteArrayDeserializer"
    }
}
```
**Provide logs (if relevant)**:
Using bundled version with 7.13.3 : 
>Pipeline error {:pipeline_id=>"main", :exception=>#<LogStash::ConfigurationError: Schema registry service doesn't respond, error: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target>

Using latest (10.8.1) version of this plugin. https + authentication: 
> [2021-07-15T13:17:49,790][ERROR][logstash.inputs.kafka    ] Unable to poll Kafka consumer {:kafka_error_message=>org.apache.kafka.common.errors.SerializationException: Error deserializing key/value for partition MYTOPIC-0 at offset 23815. If needed, please seek past the record to continue consumption., :cause=>org.apache.kafka.common.errors.SerializationException: Error retrieving Avro unknown schema for id 7}

From the errors I had with 7.11.1 I Highly suspect logstash does not pass the auth header (see this [SO article for more info](https://stackoverflow.com/questions/67586950/can-not-connect-logstash-to-confluent-schema-registry))

Using latest (10.8.1) version of this plugin, http (Haproxy for SSL termination)  + authentication **It works!**:
HaProxy Logs showing 200 http code **bolded**
>Jul 15 16:34:47 localhost haproxy[3142]: <redacted_IP>:40080 [15/Jul/2021:16:34:47.332] registry servers/srv1 0/0/65/24/89 **200** 5327 - - ---- 3/3/0/0/0 0/0 "GET /schemas/ids/7?fetchMaxId=false HTTP/1.1"

My Haproxy conf for this. 

```
frontend registry
    mode http
    bind :80
    default_backend servers


backend servers
    mode http
    server srv1 my.registry.intres:8081 ssl verify none
```

Can't test right now with an https frontend as generating a cert in my org takes a bit of time. Will update with results if I can.

Hi @robbavey . That was an interesting dive on java network debugging! 

I needed the ```-Djavax.net.debug=ssl``` to make it work. Logstash log level however didn't need to be debug. 

Here is the log I believe will be of interest to you : 

```
javax.net.ssl|DEBUG|30|kafka-input-worker-my_client_id-1|2021-07-22 13:34:08.100 UTC|Utilities.java:73|the previous server name in SNI (type=host_name (0), value=my.schema.registry) was replaced with (type=host_name (0), value=my.schema.registry)
javax.net.ssl|ERROR|30|kafka-input-worker-my_client_id-1|2021-07-22 13:34:08.127 UTC|TransportContext.java:341|Fatal (CERTIFICATE_UNKNOWN): PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target (
"throwable" : {
  sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439)
        at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306)
        at java.base/sun.security.validator.Validator.validate(Validator.java:264)
        at java.base/sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:313)
        at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:222)
        at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:129)
        at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:638)
        at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:473)
        at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:369)
        at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
        at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
        at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:421)
        at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:182)
        at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:171)
        at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1418)
        at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1324)
        at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:440)
        at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:411)
        at java.base/sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:567)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:197)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1592)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
        at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:527)
        at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:334)
        at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest(RestService.java:272)
        at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest(RestService.java:351)
        at io.confluent.kafka.schemaregistry.client.rest.RestService.getId(RestService.java:659)
        at io.confluent.kafka.schemaregistry.client.rest.RestService.getId(RestService.java:641)
        at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaByIdFromRegistry(CachedSchemaRegistryClient.java:217)
        at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaBySubjectAndId(CachedSchemaRegistryClient.java:291)
        at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaById(CachedSchemaRegistryClient.java:276)
        at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer$DeserializationContext.schemaFromRegistry(AbstractKafkaAvroDeserializer.java:273)
        at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:97)
        at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:76)
        at io.confluent.kafka.serializers.KafkaAvroDeserializer.deserialize(KafkaAvroDeserializer.java:55)
        at org.apache.kafka.common.serialization.Deserializer.deserialize(Deserializer.java:60)
        at org.apache.kafka.clients.consumer.internals.Fetcher.parseRecord(Fetcher.java:1324)
        at org.apache.kafka.clients.consumer.internals.Fetcher.access$3400(Fetcher.java:129)
        at org.apache.kafka.clients.consumer.internals.Fetcher$CompletedFetch.fetchRecords(Fetcher.java:1555)
        at org.apache.kafka.clients.consumer.internals.Fetcher$CompletedFetch.access$1700(Fetcher.java:1391)
        at org.apache.kafka.clients.consumer.internals.Fetcher.fetchRecords(Fetcher.java:683)
        at org.apache.kafka.clients.consumer.internals.Fetcher.fetchedRecords(Fetcher.java:634)
        at org.apache.kafka.clients.consumer.KafkaConsumer.pollForFetches(KafkaConsumer.java:1289)
        at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1243)
        at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1173)
        at jdk.internal.reflect.GeneratedMethodAccessor23.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:456)
        at org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:317)
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:42)
        at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_integration_minus_kafka_minus_10_dot_8_dot_1_minus_java.lib.logstash.inputs.kafka.RUBY$method$do_poll$0(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-integration-kafka-10.8.1-java/lib/logstash/inputs/kafka.rb:328)
        at usr.share.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_integration_minus_kafka_minus_10_dot_8_dot_1_minus_java.lib.logstash.inputs.kafka.RUBY$block$thread_runner$1(/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-integration-kafka-10.8.1-java/lib/logstash/inputs/kafka.rb:313)
        at org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:138)
        at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
        at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)
        at org.jruby.runtime.Block.call(Block.java:139)
        at org.jruby.RubyProc.call(RubyProc.java:318)
        at org.jruby.javasupport.Java$ProcToInterface.callProc(Java.java:1136)
        at org.jruby.javasupport.Java$ProcToInterface.access$300(Java.java:1113)
        at org.jruby.javasupport.Java$ProcToInterface$ConcreteMethod.call(Java.java:1174)
        at org.jruby.gen.InterfaceImpl1624388649.run(org/jruby/gen/InterfaceImpl1624388649.gen:13)
        at java.base/java.lang.Thread.run(Thread.java:829)
  Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
        at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
        at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)
        at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434)
        ... 62 more}

)
javax.net.ssl|DEBUG|30|kafka-input-worker-my_client_id-1|2021-07-22 13:34:08.127 UTC|SSLSocketImpl.java:1638|close the underlying socket
javax.net.ssl|DEBUG|30|kafka-input-worker-my_client_id-1|2021-07-22 13:34:08.127 UTC|SSLSocketImpl.java:1657|close the SSL connection (initiative)
```
I have a bunch of these then the classic ```Error retrieving Avro unknown schema for id```

While I have redacted both host names appearing in the line that talks about SNI and hostname replacement, I can assure you that both the initial and supposedly replaced hostname are exactly the same. 

Seems like the truststore is not applied to the http requests to the registry? Or maybe there's somethink funky going on with DNS in my org....

--- a/.gitignore
+++ b/.gitignore
@@ -11,3 +11,4 @@ build/
 .idea/
 vendor
 *.jar
+tls_repository/

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,6 @@
+## 11.2.0
+ - Added TLS truststore and keystore settings specifically to access the schema registry [#137](https://github.com/logstash-plugins/logstash-integration-kafka/pull/137)
+
 ## 11.1.0
  - Added config `group_instance_id` to use the Kafka's consumer static membership feature [#135](https://github.com/logstash-plugins/logstash-integration-kafka/pull/135)
 

--- a/docs/input-kafka.asciidoc
+++ b/docs/input-kafka.asciidoc
@@ -135,10 +135,10 @@ See the https://kafka.apache.org/{kafka_client_doc}/documentation for more detai
 | <<plugins-{type}s-{plugin}-schema_registry_key>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_proxy>> |<<uri,uri>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_secret>> |<<string,string>>|No
-| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_location>> |a valid filesystem path|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_path>> |a valid filesystem path|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_password>> |<<password,password>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_type>> |<<string,string>>, one of `["jks", "PKCS12"]`|No
-| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_location>> |a valid filesystem path|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_path>> |a valid filesystem path|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_password>> |<<password,password>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_type>> |<<string,string>>, one of `["jks", "PKCS12"]`|No
 | <<plugins-{type}s-{plugin}-schema_registry_url>> |<<uri,uri>>|No
@@ -604,8 +604,8 @@ Set the address of a forward HTTP proxy. An empty string is treated as if proxy 
 
 Set the password for basic authorization to access remote Schema Registry.
 
-[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_location"]
-===== `schema_registry_ssl_keystore_location`
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_path"]
+===== `schema_registry_ssl_keystore_path`
 
 * Value type is <<path,path>>
 * There is no default value for this setting.
@@ -628,8 +628,8 @@ If schema registry authentication is required, this setting stores the keystore 
 
 The format of the keystore file. It must be either `jks` or `PKCS12`.
 
-[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_location"]
-===== `schema_registry_ssl_truststore_location`
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_path"]
+===== `schema_registry_ssl_truststore_path`
 
 * Value type is <<path,path>>
 * There is no default value for this setting.

--- a/docs/input-kafka.asciidoc
+++ b/docs/input-kafka.asciidoc
@@ -135,6 +135,12 @@ See the https://kafka.apache.org/{kafka_client_doc}/documentation for more detai
 | <<plugins-{type}s-{plugin}-schema_registry_key>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_proxy>> |<<uri,uri>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_secret>> |<<string,string>>|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_location>> |a valid filesystem path|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_password>> |<<password,password>>|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_type>> |<<string,string>>, one of `["jks", "PKCS12"]`|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_location>> |a valid filesystem path|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_password>> |<<password,password>>|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_type>> |<<string,string>>, one of `["jks", "PKCS12"]`|No
 | <<plugins-{type}s-{plugin}-schema_registry_url>> |<<uri,uri>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_validation>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-security_protocol>> |<<string,string>>, one of `["PLAINTEXT", "SSL", "SASL_PLAINTEXT", "SASL_SSL"]`|No
@@ -597,6 +603,54 @@ Set the address of a forward HTTP proxy. An empty string is treated as if proxy 
 * There is no default value for this setting.
 
 Set the password for basic authorization to access remote Schema Registry.
+
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_location"]
+===== `schema_registry_ssl_keystore_location`
+
+* Value type is <<path,path>>
+* There is no default value for this setting.
+
+If schema registry client authentication is required, this setting stores the keystore path.
+
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_password"]
+===== `schema_registry_ssl_keystore_password`
+
+* Value type is <<password,password>>
+* There is no default value for this setting.
+
+If schema registry authentication is required, this setting stores the keystore password.
+
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_type"]
+===== `schema_registry_ssl_keystore_type`
+
+* Value type is <<string,string>>
+* There is no default value for this setting.
+
+The format of the keystore file. It must be either `jks` or `PKCS12`.
+
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_location"]
+===== `schema_registry_ssl_truststore_location`
+
+* Value type is <<path,path>>
+* There is no default value for this setting.
+
+The truststore path to validate the schema registry's certificate.
+
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_password"]
+===== `schema_registry_ssl_truststore_password`
+
+* Value type is <<password,password>>
+* There is no default value for this setting.
+
+The schema registry truststore password.
+
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_type"]
+===== `schema_registry_ssl_truststore_type`
+
+* Value type is <<string,string>>
+* There is no default value for this setting.
+
+The format of the schema registry's truststore file. It must be either `jks` or `PKCS12`.
 
 [id="plugins-{type}s-{plugin}-schema_registry_url"]
 ===== `schema_registry_url`

--- a/docs/input-kafka.asciidoc
+++ b/docs/input-kafka.asciidoc
@@ -135,10 +135,10 @@ See the https://kafka.apache.org/{kafka_client_doc}/documentation for more detai
 | <<plugins-{type}s-{plugin}-schema_registry_key>> |<<string,string>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_proxy>> |<<uri,uri>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_secret>> |<<string,string>>|No
-| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_path>> |a valid filesystem path|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_location>> |a valid filesystem path|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_password>> |<<password,password>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_keystore_type>> |<<string,string>>, one of `["jks", "PKCS12"]`|No
-| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_path>> |a valid filesystem path|No
+| <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_location>> |a valid filesystem path|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_password>> |<<password,password>>|No
 | <<plugins-{type}s-{plugin}-schema_registry_ssl_truststore_type>> |<<string,string>>, one of `["jks", "PKCS12"]`|No
 | <<plugins-{type}s-{plugin}-schema_registry_url>> |<<uri,uri>>|No
@@ -604,8 +604,8 @@ Set the address of a forward HTTP proxy. An empty string is treated as if proxy 
 
 Set the password for basic authorization to access remote Schema Registry.
 
-[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_path"]
-===== `schema_registry_ssl_keystore_path`
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_keystore_location"]
+===== `schema_registry_ssl_keystore_location`
 
 * Value type is <<path,path>>
 * There is no default value for this setting.
@@ -628,8 +628,8 @@ If schema registry authentication is required, this setting stores the keystore 
 
 The format of the keystore file. It must be either `jks` or `PKCS12`.
 
-[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_path"]
-===== `schema_registry_ssl_truststore_path`
+[id="plugins-{type}s-{plugin}-schema_registry_ssl_truststore_location"]
+===== `schema_registry_ssl_truststore_location`
 
 * Value type is <<path,path>>
 * There is no default value for this setting.

--- a/kafka_test_setup.sh
+++ b/kafka_test_setup.sh
@@ -32,7 +32,6 @@ echo "Setup Confluent Platform"
 if [ ! -e confluent-community-5.5.1-2.12.tar.gz ]; then
   echo "Confluent Platform not present locally, downloading"
   curl -s -o confluent-community-5.5.1-2.12.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
-#curl -s -o build/confluent_platform.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
 fi
 cp confluent-community-5.5.1-2.12.tar.gz build/confluent_platform.tar.gz
 mkdir build/confluent_platform && tar xzf build/confluent_platform.tar.gz -C build/confluent_platform --strip-components 1

--- a/kafka_test_setup.sh
+++ b/kafka_test_setup.sh
@@ -27,6 +27,17 @@ sleep 10
 echo "Downloading Confluent Platform"
 curl -s -o build/confluent_platform.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
 mkdir build/confluent_platform && tar xzf build/confluent_platform.tar.gz -C build/confluent_platform --strip-components 1
+
+echo "Configuring TLS on Schema registry"
+rm -Rf tls_repository
+mkdir tls_repository
+./setup_keystore_and_truststore.sh
+# configure schema-registry to handle https on 8083 port
+sed -i 's/http:\/\/0.0.0.0:8081/http:\/\/0.0.0.0:8081, https:\/\/0.0.0.0:8083/g' build/confluent_platform/etc/schema-registry/schema-registry.properties
+echo "ssl.keystore.location=`pwd`/tls_repository/schema_reg.jks" >> build/confluent_platform/etc/schema-registry/schema-registry.properties
+echo "ssl.keystore.password=changeit" >> build/confluent_platform/etc/schema-registry/schema-registry.properties
+echo "ssl.key.password=changeit" >> build/confluent_platform/etc/schema-registry/schema-registry.properties
+
 cp build/confluent_platform/etc/schema-registry/schema-registry.properties build/confluent_platform/etc/schema-registry/authed-schema-registry.properties
 echo "authentication.method=BASIC" >> build/confluent_platform/etc/schema-registry/authed-schema-registry.properties
 echo "authentication.roles=admin,developer,user,sr-user" >> build/confluent_platform/etc/schema-registry/authed-schema-registry.properties

--- a/kafka_test_setup.sh
+++ b/kafka_test_setup.sh
@@ -2,6 +2,7 @@
 # Setup Kafka and create test topics
 
 set -ex
+# check if KAFKA_VERSION env var is set
 if [ -n "${KAFKA_VERSION+1}" ]; then
   echo "KAFKA_VERSION is $KAFKA_VERSION"
 else
@@ -29,11 +30,19 @@ build/kafka/bin/kafka-server-start.sh -daemon build/kafka/config/server.properti
 sleep 10
 
 echo "Setup Confluent Platform"
-if [ ! -e confluent-community-5.5.1-2.12.tar.gz ]; then
-  echo "Confluent Platform not present locally, downloading"
-  curl -s -o confluent-community-5.5.1-2.12.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
+# check if CONFLUENT_VERSION env var is set
+if [ -n "${CONFLUENT_VERSION+1}" ]; then
+  echo "CONFLUENT_VERSION is $CONFLUENT_VERSION"
+else
+   CONFLUENT_VERSION=5.5.1
 fi
-cp confluent-community-5.5.1-2.12.tar.gz build/confluent_platform.tar.gz
+if [ ! -e confluent-community-$CONFLUENT_VERSION-2.12.tar.gz ]; then
+  echo "Confluent Platform not present locally, downloading"
+  CONFLUENT_MINOR=$(echo "$CONFLUENT_VERSION" | sed -n 's/^\([[:digit:]]*\.[[:digit:]]*\)\.[[:digit:]]*$/\1/p')
+  echo "CONFLUENT_MINOR is $CONFLUENT_MINOR"
+  curl -s -o confluent-community-$CONFLUENT_VERSION-2.12.tar.gz http://packages.confluent.io/archive/$CONFLUENT_MINOR/confluent-community-$CONFLUENT_VERSION-2.12.tar.gz
+fi
+cp confluent-community-$CONFLUENT_VERSION-2.12.tar.gz build/confluent_platform.tar.gz
 mkdir build/confluent_platform && tar xzf build/confluent_platform.tar.gz -C build/confluent_platform --strip-components 1
 
 echo "Configuring TLS on Schema registry"

--- a/kafka_test_setup.sh
+++ b/kafka_test_setup.sh
@@ -41,7 +41,11 @@ rm -Rf tls_repository
 mkdir tls_repository
 ./setup_keystore_and_truststore.sh
 # configure schema-registry to handle https on 8083 port
-sed -i 's/http:\/\/0.0.0.0:8081/http:\/\/0.0.0.0:8081, https:\/\/0.0.0.0:8083/g' build/confluent_platform/etc/schema-registry/schema-registry.properties
+if [[ "$OSTYPE" == "darwin"* ]]; then
+  sed -i '' 's/http:\/\/0.0.0.0:8081/http:\/\/0.0.0.0:8081, https:\/\/0.0.0.0:8083/g' build/confluent_platform/etc/schema-registry/schema-registry.properties
+else
+  sed -i 's/http:\/\/0.0.0.0:8081/http:\/\/0.0.0.0:8081, https:\/\/0.0.0.0:8083/g' build/confluent_platform/etc/schema-registry/schema-registry.properties
+fi
 echo "ssl.keystore.location=`pwd`/tls_repository/schema_reg.jks" >> build/confluent_platform/etc/schema-registry/schema-registry.properties
 echo "ssl.keystore.password=changeit" >> build/confluent_platform/etc/schema-registry/schema-registry.properties
 echo "ssl.key.password=changeit" >> build/confluent_platform/etc/schema-registry/schema-registry.properties

--- a/kafka_test_setup.sh
+++ b/kafka_test_setup.sh
@@ -13,8 +13,12 @@ export _JAVA_OPTIONS="-Djava.net.preferIPv4Stack=true"
 rm -rf build
 mkdir build
 
-echo "Downloading Kafka version $KAFKA_VERSION"
-curl -s -o build/kafka.tgz "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.12-$KAFKA_VERSION.tgz"
+echo "Setup Kafka version $KAFKA_VERSION"
+if [ ! -e "kafka_2.12-$KAFKA_VERSION.tgz" ]; then
+  echo "Kafka not present locally, downloading"
+  curl -s -o "kafka_2.12-$KAFKA_VERSION.tgz" "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.12-$KAFKA_VERSION.tgz"
+fi
+cp kafka_2.12-$KAFKA_VERSION.tgz build/kafka.tgz
 mkdir build/kafka && tar xzf build/kafka.tgz -C build/kafka --strip-components 1
 
 echo "Starting ZooKeeper"
@@ -24,8 +28,13 @@ echo "Starting Kafka broker"
 build/kafka/bin/kafka-server-start.sh -daemon build/kafka/config/server.properties --override advertised.host.name=127.0.0.1 --override log.dirs="${PWD}/build/kafka-logs"
 sleep 10
 
-echo "Downloading Confluent Platform"
-curl -s -o build/confluent_platform.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
+echo "Setup Confluent Platform"
+if [ ! -e confluent-community-5.5.1-2.12.tar.gz ]; then
+  echo "Confluent Platform not present locally, downloading"
+  curl -s -o confluent-community-5.5.1-2.12.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
+#curl -s -o build/confluent_platform.tar.gz http://packages.confluent.io/archive/5.5/confluent-community-5.5.1-2.12.tar.gz
+fi
+cp confluent-community-5.5.1-2.12.tar.gz build/confluent_platform.tar.gz
 mkdir build/confluent_platform && tar xzf build/confluent_platform.tar.gz -C build/confluent_platform --strip-components 1
 
 echo "Configuring TLS on Schema registry"

--- a/kafka_test_teardown.sh
+++ b/kafka_test_teardown.sh
@@ -10,3 +10,6 @@ echo "Stopping Kafka broker"
 build/kafka/bin/kafka-server-stop.sh
 echo "Stopping zookeeper"
 build/kafka/bin/zookeeper-server-stop.sh
+
+echo "Clean TLS folder"
+rm -Rf tls_repository

--- a/lib/logstash/inputs/kafka.rb
+++ b/lib/logstash/inputs/kafka.rb
@@ -460,14 +460,14 @@ class LogStash::Inputs::Kafka < LogStash::Inputs::Base
         set_trustore_keystore_config(props)
         set_sasl_config(props)
       end
-      if schema_registry_ssl_truststore_location
-        props.put('schema.registry.ssl.truststore.location', schema_registry_ssl_truststore_location)
+      if schema_registry_ssl_truststore_path
+        props.put('schema.registry.ssl.truststore.location', schema_registry_ssl_truststore_path)
         props.put('schema.registry.ssl.truststore.password', schema_registry_ssl_truststore_password.value)
         props.put('schema.registry.ssl.truststore.type', schema_registry_ssl_truststore_type)
       end
 
-      if schema_registry_ssl_keystore_location
-        props.put('schema.registry.ssl.keystore.location', schema_registry_ssl_keystore_location)
+      if schema_registry_ssl_keystore_path
+        props.put('schema.registry.ssl.keystore.location', schema_registry_ssl_keystore_path)
         props.put('schema.registry.ssl.keystore.password', schema_registry_ssl_keystore_password.value)
         props.put('schema.registry.ssl.keystore.type', schema_registry_ssl_keystore_type)
       end

--- a/lib/logstash/inputs/kafka.rb
+++ b/lib/logstash/inputs/kafka.rb
@@ -460,14 +460,14 @@ class LogStash::Inputs::Kafka < LogStash::Inputs::Base
         set_trustore_keystore_config(props)
         set_sasl_config(props)
       end
-      if schema_registry_ssl_truststore_path
-        props.put('schema.registry.ssl.truststore.location', schema_registry_ssl_truststore_path)
+      if schema_registry_ssl_truststore_location
+        props.put('schema.registry.ssl.truststore.location', schema_registry_ssl_truststore_location)
         props.put('schema.registry.ssl.truststore.password', schema_registry_ssl_truststore_password.value)
         props.put('schema.registry.ssl.truststore.type', schema_registry_ssl_truststore_type)
       end
 
-      if schema_registry_ssl_keystore_path
-        props.put('schema.registry.ssl.keystore.location', schema_registry_ssl_keystore_path)
+      if schema_registry_ssl_keystore_location
+        props.put('schema.registry.ssl.keystore.location', schema_registry_ssl_keystore_location)
         props.put('schema.registry.ssl.keystore.password', schema_registry_ssl_keystore_password.value)
         props.put('schema.registry.ssl.keystore.type', schema_registry_ssl_keystore_type)
       end

--- a/lib/logstash/inputs/kafka.rb
+++ b/lib/logstash/inputs/kafka.rb
@@ -460,6 +460,15 @@ class LogStash::Inputs::Kafka < LogStash::Inputs::Base
         set_trustore_keystore_config(props)
         set_sasl_config(props)
       end
+      if schema_registry_ssl_truststore_location
+        props.put('schema.registry.ssl.truststore.location', schema_registry_ssl_truststore_location)
+        props.put('schema.registry.ssl.truststore.password', schema_registry_ssl_truststore_password.value)
+      end
+
+      if schema_registry_ssl_keystore_location
+        props.put('schema.registry.ssl.keystore.location', schema_registry_ssl_keystore_location)
+        props.put('schema.registry.ssl.keystore.password', schema_registry_ssl_keystore_password.value)
+      end
 
       org.apache.kafka.clients.consumer.KafkaConsumer.new(props)
     rescue => e

--- a/lib/logstash/inputs/kafka.rb
+++ b/lib/logstash/inputs/kafka.rb
@@ -463,11 +463,13 @@ class LogStash::Inputs::Kafka < LogStash::Inputs::Base
       if schema_registry_ssl_truststore_location
         props.put('schema.registry.ssl.truststore.location', schema_registry_ssl_truststore_location)
         props.put('schema.registry.ssl.truststore.password', schema_registry_ssl_truststore_password.value)
+        props.put('schema.registry.ssl.truststore.type', schema_registry_ssl_truststore_type)
       end
 
       if schema_registry_ssl_keystore_location
         props.put('schema.registry.ssl.keystore.location', schema_registry_ssl_keystore_location)
         props.put('schema.registry.ssl.keystore.password', schema_registry_ssl_keystore_password.value)
+        props.put('schema.registry.ssl.keystore.type', schema_registry_ssl_keystore_type)
       end
 
       org.apache.kafka.clients.consumer.KafkaConsumer.new(props)

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -31,7 +31,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_ssl_keystore_password, :validate => :password
 
       # The keystore type
-      config :schema_registry_ssl_keystore_type, :validate => :string, :default => "jks"
+      config :schema_registry_ssl_keystore_type, :validate => ['jks', 'PKCS12'], :default => "jks"
 
       # The JKS truststore path to validate the Schema Registry's certificate.
       config :schema_registry_ssl_truststore_path, :validate => :string
@@ -40,7 +40,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_ssl_truststore_password, :validate => :password
 
       # The truststore type
-      config :schema_registry_ssl_truststore_type, :validate => :string, :default => "jks"
+      config :schema_registry_ssl_truststore_type, :validate => ['jks', 'PKCS12'], :default => "jks"
 
       # Option to skip validating the schema registry during registration. This can be useful when using
       # certificate based auth

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -27,6 +27,11 @@ module LogStash module PluginMixins module Kafka
       # Option to skip validating the schema registry during registration. This can be useful when using
       # certificate based auth
       config :schema_registry_validation, :validate => ['auto', 'skip'], :default => 'auto'
+
+      config :schema_registry_ssl_keystore_location, :validate => :string
+      config :schema_registry_ssl_keystore_password, :validate => :password
+      config :schema_registry_ssl_truststore_location, :validate => :string
+      config :schema_registry_ssl_truststore_password, :validate => :password
     end
 
     def check_schema_registry_parameters
@@ -68,6 +73,21 @@ module LogStash module PluginMixins module Kafka
       if schema_registry_key and !schema_registry_key.empty?
         options[:auth] = {:user => schema_registry_key, :password => schema_registry_secret.value}
       end
+      if schema_registry_ssl_truststore_location and !schema_registry_ssl_truststore_location.empty?
+        options[:ssl] = {} unless options.key?(:ssl)
+        options[:ssl].merge!({
+          :truststore => schema_registry_ssl_truststore_location,
+          :truststore_password => schema_registry_ssl_truststore_password.value
+        })
+      end
+      if schema_registry_ssl_keystore_location and !schema_registry_ssl_keystore_location.empty?
+        options[:ssl] = {} unless options.key? :ssl
+        options[:ssl].merge!({
+          :keystore => schema_registry_ssl_keystore_location,
+          :keystore_password => schema_registry_ssl_keystore_password.value
+        })
+      end
+
       client = Manticore::Client.new(options)
       begin
         response = client.get(@schema_registry_url.uri.to_s + '/subjects').body

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -25,7 +25,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_proxy, :validate => :uri
 
       # If schema registry client authentication is required, this setting stores the keystore path.
-      config :schema_registry_ssl_keystore_path, :validate => :string
+      config :schema_registry_ssl_keystore_location, :validate => :string
 
       # The keystore password.
       config :schema_registry_ssl_keystore_password, :validate => :password
@@ -34,7 +34,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_ssl_keystore_type, :validate => ['jks', 'PKCS12'], :default => "jks"
 
       # The JKS truststore path to validate the Schema Registry's certificate.
-      config :schema_registry_ssl_truststore_path, :validate => :string
+      config :schema_registry_ssl_truststore_location, :validate => :string
 
       # The truststore password.
       config :schema_registry_ssl_truststore_password, :validate => :password
@@ -86,15 +86,15 @@ module LogStash module PluginMixins module Kafka
       if schema_registry_key and !schema_registry_key.empty?
         options[:auth] = {:user => schema_registry_key, :password => schema_registry_secret.value}
       end
-      if schema_registry_ssl_truststore_path and !schema_registry_ssl_truststore_path.empty?
+      if schema_registry_ssl_truststore_location and !schema_registry_ssl_truststore_location.empty?
         options[:ssl] = {} unless options.key?(:ssl)
-        options[:ssl][:truststore] = schema_registry_ssl_truststore_path unless schema_registry_ssl_truststore_path.nil?
+        options[:ssl][:truststore] = schema_registry_ssl_truststore_location unless schema_registry_ssl_truststore_location.nil?
         options[:ssl][:truststore_password] = schema_registry_ssl_truststore_password.value unless schema_registry_ssl_truststore_password.nil?
         options[:ssl][:truststore_type] = schema_registry_ssl_truststore_type unless schema_registry_ssl_truststore_type.nil?
       end
-      if schema_registry_ssl_keystore_path and !schema_registry_ssl_keystore_path.empty?
+      if schema_registry_ssl_keystore_location and !schema_registry_ssl_keystore_location.empty?
         options[:ssl] = {} unless options.key? :ssl
-        options[:ssl][:keystore] = schema_registry_ssl_keystore_path unless schema_registry_ssl_keystore_path.nil?
+        options[:ssl][:keystore] = schema_registry_ssl_keystore_location unless schema_registry_ssl_keystore_location.nil?
         options[:ssl][:keystore_password] = schema_registry_ssl_keystore_password.value unless schema_registry_ssl_keystore_password.nil?
         options[:ssl][:keystore_type] = schema_registry_ssl_keystore_type unless schema_registry_ssl_keystore_type.nil?
       end

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -88,19 +88,15 @@ module LogStash module PluginMixins module Kafka
       end
       if schema_registry_ssl_truststore_path and !schema_registry_ssl_truststore_path.empty?
         options[:ssl] = {} unless options.key?(:ssl)
-        options[:ssl].merge!({
-          :truststore => schema_registry_ssl_truststore_path,
-          :truststore_password => schema_registry_ssl_truststore_password.value,
-          :truststore_type => schema_registry_ssl_truststore_type,
-        })
+        options[:ssl][:truststore] = schema_registry_ssl_truststore_path unless schema_registry_ssl_truststore_path.nil?
+        options[:ssl][:truststore_password] = schema_registry_ssl_truststore_password.value unless schema_registry_ssl_truststore_password.nil?
+        options[:ssl][:truststore_type] = schema_registry_ssl_truststore_type unless schema_registry_ssl_truststore_type.nil?
       end
       if schema_registry_ssl_keystore_path and !schema_registry_ssl_keystore_path.empty?
         options[:ssl] = {} unless options.key? :ssl
-        options[:ssl].merge!({
-          :keystore => schema_registry_ssl_keystore_path,
-          :keystore_password => schema_registry_ssl_keystore_password.value,
-          :keystore_type => schema_registry_ssl_keystore_type,
-        })
+        options[:ssl][:keystore] = schema_registry_ssl_keystore_path unless schema_registry_ssl_keystore_path.nil?
+        options[:ssl][:keystore_password] = schema_registry_ssl_keystore_password.value unless schema_registry_ssl_keystore_password.nil?
+        options[:ssl][:keystore_type] = schema_registry_ssl_keystore_type unless schema_registry_ssl_keystore_type.nil?
       end
 
       client = Manticore::Client.new(options)

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -28,10 +28,23 @@ module LogStash module PluginMixins module Kafka
       # certificate based auth
       config :schema_registry_validation, :validate => ['auto', 'skip'], :default => 'auto'
 
+      # If schema registry client authentication is required, this setting stores the keystore path.
       config :schema_registry_ssl_keystore_location, :validate => :string
+
+      # The keystore password.
       config :schema_registry_ssl_keystore_password, :validate => :password
+
+      # The keystore type
+      config :schema_registry_ssl_keystore_type, :validate => :string, :default => "JKS"
+
+      # The JKS truststore path to validate the Kafka broker's certificate.
       config :schema_registry_ssl_truststore_location, :validate => :string
+
+      # The truststore password.
       config :schema_registry_ssl_truststore_password, :validate => :password
+
+      # The truststore type
+      config :schema_registry_ssl_truststore_type, :validate => :string, :default => "JKS"
     end
 
     def check_schema_registry_parameters
@@ -77,14 +90,16 @@ module LogStash module PluginMixins module Kafka
         options[:ssl] = {} unless options.key?(:ssl)
         options[:ssl].merge!({
           :truststore => schema_registry_ssl_truststore_location,
-          :truststore_password => schema_registry_ssl_truststore_password.value
+          :truststore_password => schema_registry_ssl_truststore_password.value,
+          :truststore_type => schema_registry_ssl_truststore_type,
         })
       end
       if schema_registry_ssl_keystore_location and !schema_registry_ssl_keystore_location.empty?
         options[:ssl] = {} unless options.key? :ssl
         options[:ssl].merge!({
           :keystore => schema_registry_ssl_keystore_location,
-          :keystore_password => schema_registry_ssl_keystore_password.value
+          :keystore_password => schema_registry_ssl_keystore_password.value,
+          :keystore_type => schema_registry_ssl_keystore_type,
         })
       end
 

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -25,7 +25,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_proxy, :validate => :uri
 
       # If schema registry client authentication is required, this setting stores the keystore path.
-      config :schema_registry_ssl_keystore_location, :validate => :string
+      config :schema_registry_ssl_keystore_path, :validate => :string
 
       # The keystore password.
       config :schema_registry_ssl_keystore_password, :validate => :password
@@ -34,7 +34,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_ssl_keystore_type, :validate => :string, :default => "JKS"
 
       # The JKS truststore path to validate the Schema Registry's certificate.
-      config :schema_registry_ssl_truststore_location, :validate => :string
+      config :schema_registry_ssl_truststore_path, :validate => :string
 
       # The truststore password.
       config :schema_registry_ssl_truststore_password, :validate => :password
@@ -86,18 +86,18 @@ module LogStash module PluginMixins module Kafka
       if schema_registry_key and !schema_registry_key.empty?
         options[:auth] = {:user => schema_registry_key, :password => schema_registry_secret.value}
       end
-      if schema_registry_ssl_truststore_location and !schema_registry_ssl_truststore_location.empty?
+      if schema_registry_ssl_truststore_path and !schema_registry_ssl_truststore_path.empty?
         options[:ssl] = {} unless options.key?(:ssl)
         options[:ssl].merge!({
-          :truststore => schema_registry_ssl_truststore_location,
+          :truststore => schema_registry_ssl_truststore_path,
           :truststore_password => schema_registry_ssl_truststore_password.value,
           :truststore_type => schema_registry_ssl_truststore_type,
         })
       end
-      if schema_registry_ssl_keystore_location and !schema_registry_ssl_keystore_location.empty?
+      if schema_registry_ssl_keystore_path and !schema_registry_ssl_keystore_path.empty?
         options[:ssl] = {} unless options.key? :ssl
         options[:ssl].merge!({
-          :keystore => schema_registry_ssl_keystore_location,
+          :keystore => schema_registry_ssl_keystore_path,
           :keystore_password => schema_registry_ssl_keystore_password.value,
           :keystore_type => schema_registry_ssl_keystore_type,
         })

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -31,7 +31,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_ssl_keystore_password, :validate => :password
 
       # The keystore type
-      config :schema_registry_ssl_keystore_type, :validate => :string, :default => "JKS"
+      config :schema_registry_ssl_keystore_type, :validate => :string, :default => "jks"
 
       # The JKS truststore path to validate the Schema Registry's certificate.
       config :schema_registry_ssl_truststore_path, :validate => :string
@@ -40,7 +40,7 @@ module LogStash module PluginMixins module Kafka
       config :schema_registry_ssl_truststore_password, :validate => :password
 
       # The truststore type
-      config :schema_registry_ssl_truststore_type, :validate => :string, :default => "JKS"
+      config :schema_registry_ssl_truststore_type, :validate => :string, :default => "jks"
 
       # Option to skip validating the schema registry during registration. This can be useful when using
       # certificate based auth

--- a/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
+++ b/lib/logstash/plugin_mixins/kafka/avro_schema_registry.rb
@@ -24,10 +24,6 @@ module LogStash module PluginMixins module Kafka
       # This option permits to define a proxy to be used to reach the schema registry service instance.
       config :schema_registry_proxy, :validate => :uri
 
-      # Option to skip validating the schema registry during registration. This can be useful when using
-      # certificate based auth
-      config :schema_registry_validation, :validate => ['auto', 'skip'], :default => 'auto'
-
       # If schema registry client authentication is required, this setting stores the keystore path.
       config :schema_registry_ssl_keystore_location, :validate => :string
 
@@ -37,7 +33,7 @@ module LogStash module PluginMixins module Kafka
       # The keystore type
       config :schema_registry_ssl_keystore_type, :validate => :string, :default => "JKS"
 
-      # The JKS truststore path to validate the Kafka broker's certificate.
+      # The JKS truststore path to validate the Schema Registry's certificate.
       config :schema_registry_ssl_truststore_location, :validate => :string
 
       # The truststore password.
@@ -45,6 +41,10 @@ module LogStash module PluginMixins module Kafka
 
       # The truststore type
       config :schema_registry_ssl_truststore_type, :validate => :string, :default => "JKS"
+
+      # Option to skip validating the schema registry during registration. This can be useful when using
+      # certificate based auth
+      config :schema_registry_validation, :validate => ['auto', 'skip'], :default => 'auto'
     end
 
     def check_schema_registry_parameters

--- a/logstash-integration-kafka.gemspec
+++ b/logstash-integration-kafka.gemspec
@@ -1,6 +1,6 @@
 Gem::Specification.new do |s|
   s.name            = 'logstash-integration-kafka'
-  s.version         = '11.1.0'
+  s.version         = '11.2.0'
   s.licenses        = ['Apache-2.0']
   s.summary         = "Integration with Kafka - input and output plugins"
   s.description     = "This gem is a Logstash plugin required to be installed on top of the Logstash core pipeline "+

--- a/setup_keystore_and_truststore.sh
+++ b/setup_keystore_and_truststore.sh
@@ -1,0 +1,12 @@
+#!/bin/bash
+# Setup Schema Registry keystore and Kafka's schema registry client's truststore
+set -ex
+
+echo "Generating schema registry key store"
+keytool -genkey -alias schema_reg -keyalg RSA -keystore tls_repository/schema_reg.jks -keypass changeit -storepass changeit  -validity 365  -keysize 2048 -dname "CN=localhost, OU=John Doe, O=Acme Inc, L=Unknown, ST=Unknown, C=IT"
+
+echo "Exporting schema registry certificate"
+keytool -exportcert -rfc -keystore tls_repository/schema_reg.jks -storepass changeit -alias schema_reg -file tls_repository/schema_reg_certificate.pem
+
+echo "Creating client's truststore and importing schema registry's certificate"
+keytool -import -trustcacerts -file tls_repository/schema_reg_certificate.pem -keypass changeit -storepass changeit -keystore tls_repository/clienttruststore.jks -noprompt

--- a/spec/integration/inputs/kafka_spec.rb
+++ b/spec/integration/inputs/kafka_spec.rb
@@ -557,7 +557,7 @@ describe "Deserializing with the schema registry", :integration => true do
     let(:subject_url) { "#{proto}://localhost:#{port}/subjects" }
     let(:plain_config)  { base_config.merge!({
       'schema_registry_url' => "#{proto}://localhost:#{port}",
-      'schema_registry_ssl_truststore_location' => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+      'schema_registry_ssl_truststore_path' => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
       'schema_registry_ssl_truststore_password' => 'changeit',
     }) }
 
@@ -585,7 +585,7 @@ describe "Deserializing with the schema registry", :integration => true do
     let(:tls_base_config) do
       if tls
         base_config.merge({
-          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+          'schema_registry_ssl_truststore_path' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
           'schema_registry_ssl_truststore_password' => 'changeit',
         })
       else

--- a/spec/integration/inputs/kafka_spec.rb
+++ b/spec/integration/inputs/kafka_spec.rb
@@ -353,10 +353,13 @@ describe "schema registry connection options" do
   end
 end
 
-def save_avro_schema_to_schema_registry(schema_file, subject_name)
+def save_avro_schema_to_schema_registry(schema_file, subject_name, proto = 'http', port = 8081, manticore_options = {})
   raw_schema = File.readlines(schema_file).map(&:chomp).join
   raw_schema_quoted = raw_schema.gsub('"', '\"')
-  response = Manticore.post("http://localhost:8081/subjects/#{subject_name}/versions",
+
+  client = Manticore::Client.new(manticore_options)
+
+  response = client.post("#{proto}://localhost:#{port}/subjects/#{subject_name}/versions",
           body: '{"schema": "' + raw_schema_quoted + '"}',
           headers: {"Content-Type" => "application/vnd.schemaregistry.v1+json"})
   response
@@ -378,8 +381,17 @@ def startup_schema_registry(schema_registry, auth=false)
   end
 end
 
-describe "Schema registry API", :integration => true do
-  schema_registry = Manticore::Client.new
+shared_examples 'it has endpoints available to' do |tls|
+  let(:port) { tls ? 8083 : 8081 }
+  let(:proto) { tls ? 'https' : 'http' }
+
+  manticore_options = {
+    :ssl => {
+      :truststore => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+      :truststore_password => "changeit"
+    }
+  }
+  schema_registry = Manticore::Client.new(manticore_options)
 
   before(:all) do
     startup_schema_registry(schema_registry)
@@ -391,27 +403,38 @@ describe "Schema registry API", :integration => true do
 
   context 'listing subject on clean instance' do
     it "should return an empty set" do
-      subjects = JSON.parse schema_registry.get('http://localhost:8081/subjects').body
+      subjects = JSON.parse schema_registry.get("#{proto}://localhost:#{port}/subjects").body
       expect( subjects ).to be_empty
     end
   end
 
   context 'send a schema definition' do
     it "save the definition" do
-      response = save_avro_schema_to_schema_registry(File.join(Dir.pwd, "spec", "unit", "inputs", "avro_schema_fixture_payment.asvc"), "schema_test_1")
+      response = save_avro_schema_to_schema_registry(File.join(Dir.pwd, "spec", "unit", "inputs", "avro_schema_fixture_payment.asvc"), "schema_test_1", proto, port, manticore_options)
       expect( response.code ).to be(200)
       delete_remote_schema(schema_registry, "schema_test_1")
     end
 
     it "delete the schema just added" do
-      response = save_avro_schema_to_schema_registry(File.join(Dir.pwd, "spec", "unit", "inputs", "avro_schema_fixture_payment.asvc"), "schema_test_1")
+      response = save_avro_schema_to_schema_registry(File.join(Dir.pwd, "spec", "unit", "inputs", "avro_schema_fixture_payment.asvc"), "schema_test_1", proto, port, manticore_options)
       expect( response.code ).to be(200)
 
-      expect( schema_registry.delete('http://localhost:8081/subjects/schema_test_1?permanent=false').code ).to be(200)
+      expect( schema_registry.delete("#{proto}://localhost:#{port}/subjects/schema_test_1?permanent=false").code ).to be(200)
       sleep(1)
-      subjects = JSON.parse schema_registry.get('http://localhost:8081/subjects').body
+      subjects = JSON.parse schema_registry.get("#{proto}://localhost:#{port}/subjects").body
       expect( subjects ).to be_empty
     end
+  end
+end
+
+describe "Schema registry API", :integration => true do
+
+  context "when exposed with HTTPS" do
+    it_behaves_like 'it has endpoints available to', true
+  end
+
+  context "when exposed with plain HTTP" do
+    it_behaves_like 'it has endpoints available to', false
   end
 end
 

--- a/spec/integration/inputs/kafka_spec.rb
+++ b/spec/integration/inputs/kafka_spec.rb
@@ -443,7 +443,13 @@ def shutdown_schema_registry
 end
 
 describe "Deserializing with the schema registry", :integration => true do
-  schema_registry = Manticore::Client.new
+  manticore_options = {
+    :ssl => {
+      :truststore => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+      :truststore_password => "changeit"
+    }
+  }
+  schema_registry = Manticore::Client.new(manticore_options)
 
   shared_examples 'it reads from a topic using a schema registry' do |with_auth|
 
@@ -542,26 +548,41 @@ describe "Deserializing with the schema registry", :integration => true do
     end
   end
 
-  context 'with an unauthed schema registry' do
+  shared_examples 'with an unauthed schema registry' do |tls|
+    let(:port) { tls ? 8083 : 8081 }
+    let(:proto) { tls ? 'https' : 'http' }
+
     let(:auth) { false }
     let(:avro_topic_name) { "topic_avro" }
-    let(:subject_url) { "http://localhost:8081/subjects" }
-    let(:plain_config)  { base_config.merge!({'schema_registry_url' => "http://localhost:8081"}) }
+    let(:subject_url) { "#{proto}://localhost:#{port}/subjects" }
+    let(:plain_config)  { base_config.merge!({'schema_registry_url' => "#{proto}://localhost:#{port}"}) }
 
     it_behaves_like 'it reads from a topic using a schema registry', false
   end
 
-  context 'with an authed schema registry' do
+  context 'with an unauthed schema registry' do
+    context "accessed through HTTPS" do
+      it_behaves_like 'with an unauthed schema registry', true
+    end
+
+    context "accessed through HTTPS" do
+      it_behaves_like 'with an unauthed schema registry', false
+    end
+  end
+
+  shared_examples 'with an authed schema registry' do |tls|
+    let(:port) { tls ? 8083 : 8081 }
+    let(:proto) { tls ? 'https' : 'http' }
     let(:auth) { true }
     let(:user) { "barney" }
     let(:password) { "changeme" }
     let(:avro_topic_name) { "topic_avro_auth" }
-    let(:subject_url) { "http://#{user}:#{password}@localhost:8081/subjects" }
+    let(:subject_url) { "#{proto}://#{user}:#{password}@localhost:#{port}/subjects" }
 
     context 'using schema_registry_key' do
       let(:plain_config) do
         base_config.merge!({
-          'schema_registry_url' => "http://localhost:8081",
+          'schema_registry_url' => "#{proto}://localhost:#{port}",
           'schema_registry_key' => user,
           'schema_registry_secret' => password
         })
@@ -573,11 +594,21 @@ describe "Deserializing with the schema registry", :integration => true do
     context 'using schema_registry_url' do
       let(:plain_config) do
         base_config.merge!({
-          'schema_registry_url' => "http://#{user}:#{password}@localhost:8081"
+          'schema_registry_url' => "#{proto}://#{user}:#{password}@localhost:#{port}"
         })
       end
 
       it_behaves_like 'it reads from a topic using a schema registry', true
+    end
+  end
+
+  context 'with an authed schema registry' do
+    context "accessed through HTTPS" do
+      it_behaves_like 'with an authed schema registry', true
+    end
+
+    context "accessed through HTTPS" do
+      it_behaves_like 'with an authed schema registry', false
     end
   end
 end

--- a/spec/integration/inputs/kafka_spec.rb
+++ b/spec/integration/inputs/kafka_spec.rb
@@ -557,7 +557,7 @@ describe "Deserializing with the schema registry", :integration => true do
     let(:subject_url) { "#{proto}://localhost:#{port}/subjects" }
     let(:plain_config)  { base_config.merge!({
       'schema_registry_url' => "#{proto}://localhost:#{port}",
-      'schema_registry_ssl_truststore_path' => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+      'schema_registry_ssl_truststore_location' => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
       'schema_registry_ssl_truststore_password' => 'changeit',
     }) }
 
@@ -585,7 +585,7 @@ describe "Deserializing with the schema registry", :integration => true do
     let(:tls_base_config) do
       if tls
         base_config.merge({
-          'schema_registry_ssl_truststore_path' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
           'schema_registry_ssl_truststore_password' => 'changeit',
         })
       else

--- a/spec/integration/inputs/kafka_spec.rb
+++ b/spec/integration/inputs/kafka_spec.rb
@@ -582,15 +582,23 @@ describe "Deserializing with the schema registry", :integration => true do
     let(:password) { "changeme" }
     let(:avro_topic_name) { "topic_avro_auth" }
     let(:subject_url) { "#{proto}://#{user}:#{password}@localhost:#{port}/subjects" }
+    let(:tls_base_config) do
+      if tls
+        base_config.merge({
+          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+          'schema_registry_ssl_truststore_password' => 'changeit',
+        })
+      else
+        base_config
+      end
+    end
 
     context 'using schema_registry_key' do
       let(:plain_config) do
-        base_config.merge!({
+        tls_base_config.merge!({
           'schema_registry_url' => "#{proto}://localhost:#{port}",
           'schema_registry_key' => user,
           'schema_registry_secret' => password,
-          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
-          'schema_registry_ssl_truststore_password' => 'changeit',
         })
       end
 
@@ -599,10 +607,8 @@ describe "Deserializing with the schema registry", :integration => true do
 
     context 'using schema_registry_url' do
       let(:plain_config) do
-        base_config.merge!({
+        tls_base_config.merge!({
           'schema_registry_url' => "#{proto}://#{user}:#{password}@localhost:#{port}",
-          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
-          'schema_registry_ssl_truststore_password' => 'changeit',
         })
       end
 

--- a/spec/integration/inputs/kafka_spec.rb
+++ b/spec/integration/inputs/kafka_spec.rb
@@ -555,7 +555,11 @@ describe "Deserializing with the schema registry", :integration => true do
     let(:auth) { false }
     let(:avro_topic_name) { "topic_avro" }
     let(:subject_url) { "#{proto}://localhost:#{port}/subjects" }
-    let(:plain_config)  { base_config.merge!({'schema_registry_url' => "#{proto}://localhost:#{port}"}) }
+    let(:plain_config)  { base_config.merge!({
+      'schema_registry_url' => "#{proto}://localhost:#{port}",
+      'schema_registry_ssl_truststore_location' => File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+      'schema_registry_ssl_truststore_password' => 'changeit',
+    }) }
 
     it_behaves_like 'it reads from a topic using a schema registry', false
   end
@@ -584,7 +588,9 @@ describe "Deserializing with the schema registry", :integration => true do
         base_config.merge!({
           'schema_registry_url' => "#{proto}://localhost:#{port}",
           'schema_registry_key' => user,
-          'schema_registry_secret' => password
+          'schema_registry_secret' => password,
+          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+          'schema_registry_ssl_truststore_password' => 'changeit',
         })
       end
 
@@ -594,7 +600,9 @@ describe "Deserializing with the schema registry", :integration => true do
     context 'using schema_registry_url' do
       let(:plain_config) do
         base_config.merge!({
-          'schema_registry_url' => "#{proto}://#{user}:#{password}@localhost:#{port}"
+          'schema_registry_url' => "#{proto}://#{user}:#{password}@localhost:#{port}",
+          'schema_registry_ssl_truststore_location' => ::File.join(Dir.pwd, "tls_repository/clienttruststore.jks"),
+          'schema_registry_ssl_truststore_password' => 'changeit',
         })
       end
 

--- a/stop_schema_registry.sh
+++ b/stop_schema_registry.sh
@@ -2,6 +2,6 @@
 # Setup Kafka and create test topics
 set -ex
 
-echo "Stoppping SchemaRegistry"
+echo "Stopping SchemaRegistry"
 build/confluent_platform/bin/schema-registry-stop
 sleep 5

